csrc/multidevice/lower_communication.cpp

```diff
@@ -0,0 +1,324 @@
+// see 'CreateParamsForGatherScatter'
+inline at::Tensor createDummyTensor(at::Tensor reference) {
+  return at::empty_like(reference, reference.options());
+}
+
+// Utility function used for setting up a scatter or gather communication
+// params. Since most  of the steps are somewhat similar/opposite in those
+// cases, we gathered the two implementations into one function. The argument
+// "is_scatter" allows to discriminate between scatter and gather
+CommParams CreateParamsForGatherScatter(
+    DeviceIdxType device_index,
+    DeviceIdxType root,
+    DeviceMesh mesh, // is_scatter? receivers : senders
+    at::Tensor root_buf, // is_scatter? input buf : output buf
+    at::Tensor buf, // is_scatter? output buf : input buf
+    bool is_scatter) {
+  CommParams params;
+  params.root = root;
+  params.team = mesh.vector();
+  bool is_root_in_mesh = mesh.has(root);
+  if (!is_root_in_mesh) {
+    params.team.push_back(root);
+  }
+
+  if (mesh.has(device_index)) {
+    auto sliced_buf =
+        buf.index({static_cast<int>(mesh.findIndex(device_index)), "..."});
+    ((is_scatter) ? params.dst_bufs : params.src_bufs) = {sliced_buf};
+  }
+
+  if (device_index == root) {
+    for (auto i : c10::irange(mesh.vector().size())) {
+      ((is_scatter) ? params.src_bufs : params.dst_bufs)
+          .push_back(root_buf.index({static_cast<int>(i), "..."}));
+    }
+    // The scatter/gather semantics imposes the root to be both
+    // sender and receiver. If the root is not in the mesh, we thus
+    // have to artificially make it send and receive a dummy buffer
+    // Since it is an "inplace" operation, this should not cause any overhead
+    if (!is_root_in_mesh) {
+      at::Tensor dummy = createDummyTensor(root_buf.index({0, "..."}));
+      params.src_bufs.push_back(dummy);
+      params.dst_bufs.push_back(dummy);
+    }
+  }
+  return params;
+}
+
+// Adds one or zero Scatter communication to the vector 'comms'
+void lowerToScatter(
+    DeviceIdxType device_index,
+    DeviceMesh sender_mesh,
+    DeviceMesh receiver_mesh,
+    at::Tensor input_tensor,
+    at::Tensor output_tensor,
+    std::vector<std::shared_ptr<Communication>>& comms) {
+  // we arbitrarily choose the first device of the sender mesh to be the root
+  auto root = sender_mesh.vector().at(0);
+  if (!isDeviceInvolved(device_index, root, receiver_mesh)) {
+    return;
+  }
+  auto params = CreateParamsForGatherScatter(
+      device_index, root, receiver_mesh, input_tensor, output_tensor, true);
+  comms.push_back(std::make_shared<Scatter>(std::move(params)));
+}
+
+// Adds one or zero Gather communication to the vector 'comms'
+void lowerToGather(
+    DeviceIdxType device_index,
+    DeviceMesh sender_mesh,
+    DeviceMesh receiver_mesh,
+    at::Tensor input_tensor,
+    at::Tensor output_tensor,
+    std::vector<std::shared_ptr<Communication>>& comms) {
+  // we create as many 'Gathers' as there are devices in the receiver mesh
```

Comment:
Thanks. That makes sense. Could you add some notes here? I now see that the comments you already have actual indicate what you described above, but it would be also helpful to have that as a code comment as well

Reply:
I added a comment here